File: ML-Practicas/P2 - Linear Regresion/Maquina de Soporte Vecorial/svr.R
#Plantilla de Regresión

dataset=read.csv('Position_Salaries.csv')
dataset=dataset[,2:3] #Solo ocupamos las columnas 2 y 3

#dividir los datos en conjuntos de entrenamiento y test
#install.packages("caTools")
library(caTools)
#set.seed(123) #Semilla para aleatorios

#Dividir en data set de entrenamiento y test
#En este caso no se dividen porque hay pocos datos

# split=sample.split(dataset$Salary,SplitRatio =2/3)
# training_set=subset(dataset,split==TRUE)
# testing_set=subset(dataset,split==FALSE)

# Escalado de valores
# training_set[,2:3]=scale(training_set[,2:3])
# testing_set[,2:3]=scale(testing_set[,2:3])

#Ajustar Modelo re Regresión con el Conunto de datos

#install.packages("e1071") #Libreria para modelos SVR
regression=svm(formula=Salary ~., 
               data = dataset,
               type="eps-regression",
               kernel="radial")

#Predicción de resultados de la regresión
y_pred=predict(regression,newdata = data.frame(Level=6.5))



#Visualización  del modelo con regresion
#install.packages("ggplot2")
library(ggplot2)

#x_grid=seq(min(dataset$Level),max(dataset$Level),0.1) #Suavisando la curva graficando más valores

ggplot()+
  geom_point(aes(x=dataset$Level,y=dataset$Salary),
             color="red")+
  geom_line(aes(x=dataset$Level,y=predict(regression,newdata = data.frame(Level=dataset$Level))),
            color="blue")+
  ggtitle("Predicción (Modelo de regresión SVR)")+
  xlab("Etiqueta Eje X")+
  ylab("Etiqueta Eje Y")

File: ML-Practicas/P2 - Linear Regresion/Maquina de Soporte Vecorial/svr.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jun 21 16:09:19 2019
@author: adrian.perez
"""
#Modelo de soporte vectorial

#Preprocesado 
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
#importar el dataset
dataset = pd.read_csv("Position_salaries.csv")
X=dataset.iloc[:,1:2].values #EStablece las variables independientes
y=dataset.iloc[:, 2].values #establecer variable dependiente

#Dividir el dataset en conjuto de entrenamiento y conjunto de test
"""
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=1/3,random_state=0)
"""
#Escalado de vaiables
from sklearn.preprocessing import StandardScaler
sc_X=StandardScaler()
sc_y=StandardScaler()
X=sc_X.fit_transform(X)
y=sc_y.fit_transform(y.reshape(-1,1))


# Ajustar regresión con el data set
# Crear modelo de regression
from sklearn.svm import SVR
regression=SVR(kernel="rbf")
regression.fit(X,y)

#Predicción de los modelos de SVR

y_pred=regression.predict(sc_X.transform([[6.5]]))
#y_pred=regression.predict(X_test)

#aplicamos el escalado inverso
y_pred=sc_y.inverse_transform(y_pred)

#visualizacipon de los resultados del  modelo SVR

X_grid=np.arange(min(X),max(X),0.1) # Ajustar los valores de X para un mejro calidad en la grafica(unicamente)
X_grid=X_grid.reshape(len(X_grid),1)
plt.scatter(X,y,color="red")
#plt.plot(X,lin_reg_2.predict(X_poly),color="blue")
plt.plot(X_grid,regression.predict(X_grid),color="blue")
plt.title("Modelo de regresion SVR")
plt.xlabel("")
plt.ylabel("")
plt.show()

#revertimos es escalado para graficar los valores reales
X=sc_X.inverse_transform(X)
y=sc_y.inverse_transform(y)

X_grid=np.arange(min(X),max(X),0.1) # Ajustar los valores de X para un mejro calidad en la grafica(unicamente)
X_grid=X_grid.reshape(len(X_grid),1)
plt.scatter(X,y,color="red")
#plt.plot(X,lin_reg_2.predict(X_poly),color="blue")
#plt.plot(X_grid,regression.predict(X_grid),color="blue")
plt.title("Modelo de regresion SVR")
plt.xlabel("")
plt.ylabel("")
plt.show()
